File: android/liveness/src/main/java/com/liveness/detection/LivenessOvalOverlayView.kt
package com.liveness.detection

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import kotlin.math.max
import kotlin.math.min

/**
 * Overlay: dark mask with oval/circle cutout, progress ring, optional step dots.
 *
 * Shape, colours, stroke width, line cap and minimum diameter are configurable via
 * [applyStyle] using the theme fields in [LivenessConfig]. When [showInstructions]
 * is false the step dots are hidden (text labels are owned by the parent view).
 */
class LivenessOvalOverlayView @JvmOverloads constructor(
  context: Context,
  attrs: AttributeSet? = null,
  defStyleAttr: Int = 0,
) : View(context, attrs, defStyleAttr) {

  companion object {
    private const val STEP_COUNT = 5
    // Max radii in dp — same semantics as the web SDK's OVAL_BASE / CIRCLE_BASE.
    private const val OVAL_MAX_W_DP = 270f
    private const val OVAL_MAX_H_DP = 360f
    private const val CIRCLE_MAX_DP = 270f
  }

  private val density = context.resources.displayMetrics.density
  private val ovalRect = RectF()

  // Paints with default web-parity values; replaced via applyStyle when config is set.
  private val darkPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = 0xD1000000.toInt()
    style = Paint.Style.FILL
  }
  private val darkPaintOutOfOval = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = 0x8C000000.toInt()
    style = Paint.Style.FILL
  }
  private val trackPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = Color.argb(38, 255, 255, 255)
    style = Paint.Style.STROKE
    strokeWidth = 3.5f * density
  }
  private val progressPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = Color.parseColor("#12c95c")
    style = Paint.Style.STROKE
    strokeWidth = 3.5f * density
    strokeCap = Paint.Cap.ROUND
  }
  private val progressPaintOutOfOval = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = Color.parseColor("#ff3b3b")
    style = Paint.Style.STROKE
    strokeWidth = 3.5f * density
    strokeCap = Paint.Cap.ROUND
  }
  private val dotPaintInactive = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = Color.argb(51, 255, 255, 255)
    style = Paint.Style.FILL
  }
  private val dotPaintActive = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = Color.parseColor("#12c95c")
    style = Paint.Style.FILL
  }
  private val dotPaintDone = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = Color.parseColor("#12c95c")
    style = Paint.Style.FILL
    alpha = 128
  }

  private var shape: String = "oval"
  private var minSizePx: Float = 220f * density
  private var showDots: Boolean = true
  private var ovalTopFraction: Float = 0.40f

  var faceInOval = true
    set(value) {
      if (field != value) { field = value; invalidate() }
    }
  var completedSteps = 0
    set(value) {
      val v = value.coerceIn(0, STEP_COUNT)
      if (field != v) { field = v; invalidate() }
    }
  var activeStepIndex = 0
    set(value) {
      val v = value.coerceIn(0, STEP_COUNT - 1)
      if (field != v) { field = v; invalidate() }
    }

  fun setFaceInOval(inside: Boolean) { faceInOval = inside }
  fun setProgress(completed: Int) { completedSteps = completed }
  fun setStepDots(activeIndex: Int) { activeStepIndex = activeIndex }

  /** Apply theme, shape and minimum size from [LivenessConfig]. */
  fun applyStyle(c: LivenessConfig) {
    shape = if (c.shape.equals("circle", ignoreCase = true)) "circle" else "oval"
    minSizePx = max(0f, c.minSizeDp) * density
    showDots = c.showInstructions
    // Place the cutout slightly above centre when there's room below for labels; otherwise centre it.
    ovalTopFraction = if (c.showInstructions) 0.40f else 0.50f

    darkPaint.color = c.overlayColor
    darkPaintOutOfOval.color = c.overlayErrorColor

    val cap = when (c.progressLineCap.lowercase()) {
      "square" -> Paint.Cap.SQUARE
      "butt"   -> Paint.Cap.BUTT
      else     -> Paint.Cap.ROUND
    }
    val strokePx = c.progressWidthDp * density
    progressPaint.color = c.progressColor
    progressPaint.strokeWidth = strokePx
    progressPaint.strokeCap = cap
    progressPaintOutOfOval.color = c.progressErrorColor
    progressPaintOutOfOval.strokeWidth = strokePx
    progressPaintOutOfOval.strokeCap = cap
    trackPaint.strokeWidth = strokePx

    dotPaintActive.color = c.progressColor
    dotPaintDone.color = c.progressColor
    dotPaintDone.alpha = 128

    invalidate()
  }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    val w = width.toFloat()
    val h = height.toFloat()

    // Base sizing matches the web SDK: width = min(45vmin-equivalent, max-dp); height derives.
    val ovalW: Float
    val ovalH: Float
    if (shape == "circle") {
      val side = max(minSizePx, min(min(w * 0.72f, h * 0.55f), CIRCLE_MAX_DP * density))
      ovalW = side
      ovalH = side
    } else {
      val baseW = max(minSizePx, min(w * 0.72f, OVAL_MAX_W_DP * density))
      val baseH = max(minSizePx * (OVAL_MAX_H_DP / OVAL_MAX_W_DP), min(h * 0.55f, OVAL_MAX_H_DP * density))
      // Keep oval aspect roughly 270:360 unless screen forces shrinking
      val aspectH = baseW * (OVAL_MAX_H_DP / OVAL_MAX_W_DP)
      ovalW = baseW
      ovalH = min(baseH, aspectH)
    }

    val cx = w / 2f
    val cy = h * ovalTopFraction
    ovalRect.set(cx - ovalW / 2, cy - ovalH / 2, cx + ovalW / 2, cy + ovalH / 2)

    val overlayPaint = if (faceInOval) darkPaint else darkPaintOutOfOval
    val path = Path().apply { addRect(0f, 0f, w, h, Path.Direction.CW) }
    val cutoutPath = Path().apply { addOval(ovalRect, Path.Direction.CW) }
    path.op(cutoutPath, Path.Op.DIFFERENCE)
    canvas.drawPath(path, overlayPaint)

    canvas.drawOval(ovalRect, trackPaint)
    val progressPaintToUse = if (faceInOval) progressPaint else progressPaintOutOfOval
    val sweep = 360f * (completedSteps.toFloat() / STEP_COUNT)
    canvas.drawArc(ovalRect, -90f, sweep, false, progressPaintToUse)

    if (!showDots) return

    val dotRadius = 3.5f * density
    val dotGap = 8f * density
    val totalDotsWidth = STEP_COUNT * (dotRadius * 2) + (STEP_COUNT - 1) * dotGap
    var dotLeft = cx - totalDotsWidth / 2f + dotRadius
    val dotY = cy + ovalH / 2f + 20f * density
    for (i in 0 until STEP_COUNT) {
      val p = when {
        i < activeStepIndex -> dotPaintDone
        i == activeStepIndex -> dotPaintActive
        else -> dotPaintInactive
      }
      canvas.drawCircle(dotLeft, dotY, dotRadius, p)
      dotLeft += dotRadius * 2 + dotGap
    }
  }
}

File: ios/LivenessDetection/LivenessOvalOverlayView.swift
import UIKit

/// Overlay: dark mask with oval/circle cutout, progress ring, optional step dots.
///
/// Shape, colours, stroke width, line cap and minimum diameter are configurable via
/// `applyStyle(_:)` using the theme fields in `LivenessConfig`. When `showInstructions`
/// is false the step dots are hidden (text labels are owned by the parent view).
public final class LivenessOvalOverlayView: UIView {
  private static let stepCount = 5
  // Max radii (pt) — mirrors web SDK's OVAL_BASE / CIRCLE_BASE.
  private static let ovalMaxW: CGFloat = 270
  private static let ovalMaxH: CGFloat = 360
  private static let circleMax: CGFloat = 270

  private let darkLayer = CALayer()
  private let ovalMaskLayer = CAShapeLayer()
  private let trackLayer = CAShapeLayer()
  private let progressLayer = CAShapeLayer()
  private var dotLayers: [CAShapeLayer] = []

  private var faceInOval = true
  private var completedSteps = 0
  private var activeStepIndex = 0

  // Styling — populated by applyStyle, with web-parity defaults.
  private var shape: String = "oval"
  private var minSize: CGFloat = 220
  private var showDots: Bool = true
  private var ovalTopFraction: CGFloat = 0.40
  private var progressColor: UIColor = UIColor(red: 18/255, green: 201/255, blue: 92/255, alpha: 1)
  private var progressErrorColor: UIColor = UIColor(red: 1, green: 59/255, blue: 59/255, alpha: 1)
  private var progressWidth: CGFloat = 3.5
  private var progressLineCap: CAShapeLayerLineCap = .round
  private var overlayColor: UIColor = UIColor(white: 0, alpha: 0.82)
  private var overlayErrorColor: UIColor = UIColor(red: 180/255, green: 0, blue: 0, alpha: 0.55)

  override public init(frame: CGRect) {
    super.init(frame: frame)
    setupLayers()
  }

  required init?(coder: NSCoder) {
    super.init(coder: coder)
    setupLayers()
  }

  private func setupLayers() {
    darkLayer.backgroundColor = overlayColor.cgColor
    layer.addSublayer(darkLayer)
    ovalMaskLayer.fillRule = .evenOdd
    darkLayer.mask = ovalMaskLayer
    trackLayer.fillColor = nil
    trackLayer.strokeColor = UIColor.white.withAlphaComponent(0.15).cgColor
    trackLayer.lineWidth = progressWidth
    layer.addSublayer(trackLayer)
    progressLayer.fillColor = nil
    progressLayer.lineWidth = progressWidth
    progressLayer.lineCap = progressLineCap
    progressLayer.strokeColor = progressColor.cgColor
    layer.addSublayer(progressLayer)
    for _ in 0..<Self.stepCount {
      let dot = CAShapeLayer()
      dot.path = UIBezierPath(ovalIn: CGRect(x: 0, y: 0, width: 7, height: 7)).cgPath
      layer.addSublayer(dot)
      dotLayers.append(dot)
    }
  }

  public func applyStyle(_ c: LivenessConfig) {
    shape = c.shape.lowercased() == "circle" ? "circle" : "oval"
    minSize = max(0, c.minSize)
    showDots = c.showInstructions
    ovalTopFraction = c.showInstructions ? 0.40 : 0.50

    progressColor = c.progressColor
    progressErrorColor = c.progressErrorColor
    progressWidth = c.progressWidth
    overlayColor = c.overlayColor
    overlayErrorColor = c.overlayErrorColor
    switch c.progressLineCap.lowercased() {
    case "square": progressLineCap = .square
    case "butt":   progressLineCap = .butt
    default:       progressLineCap = .round
    }

    trackLayer.lineWidth = progressWidth
    progressLayer.lineWidth = progressWidth
    progressLayer.lineCap = progressLineCap
    progressLayer.strokeColor = (faceInOval ? progressColor : progressErrorColor).cgColor
    darkLayer.backgroundColor = (faceInOval ? overlayColor : overlayErrorColor).cgColor

    for dot in dotLayers { dot.isHidden = !showDots }

    setNeedsLayout()
  }

  public func setFaceInOval(_ inside: Bool) {
    guard faceInOval != inside else { return }
    faceInOval = inside
    darkLayer.backgroundColor = (inside ? overlayColor : overlayErrorColor).cgColor
    progressLayer.strokeColor = (inside ? progressColor : progressErrorColor).cgColor
  }

  public func setProgress(_ completed: Int) {
    guard completedSteps != completed else { return }
    completedSteps = min(max(completed, 0), Self.stepCount)
    setNeedsLayout()
  }

  public func setStepDots(activeIndex: Int) {
    guard activeStepIndex != activeIndex else { return }
    activeStepIndex = min(max(activeIndex, 0), Self.stepCount - 1)
    setNeedsLayout()
  }

  override public func layoutSubviews() {
    super.layoutSubviews()
    let w = bounds.width
    let h = bounds.height

    let ovalW: CGFloat
    let ovalH: CGFloat
    if shape == "circle" {
      let side = max(minSize, min(min(w * 0.72, h * 0.55), Self.circleMax))
      ovalW = side
      ovalH = side
    } else {
      let baseW = max(minSize, min(w * 0.72, Self.ovalMaxW))
      let aspectH = baseW * (Self.ovalMaxH / Self.ovalMaxW)
      let baseH = max(minSize * (Self.ovalMaxH / Self.ovalMaxW), min(h * 0.55, Self.ovalMaxH))
      ovalW = baseW
      ovalH = min(baseH, aspectH)
    }

    let cx = w / 2
    let cy = h * ovalTopFraction
    let ovalRect = CGRect(x: cx - ovalW/2, y: cy - ovalH/2, width: ovalW, height: ovalH)

    darkLayer.frame = bounds
    ovalMaskLayer.frame = bounds
    let fullPath = UIBezierPath(rect: bounds)
    fullPath.append(UIBezierPath(ovalIn: ovalRect))
    ovalMaskLayer.path = fullPath.cgPath

    trackLayer.frame = bounds
    trackLayer.path = UIBezierPath(ovalIn: ovalRect).cgPath

    // Approximate progress sweep along the ellipse perimeter
    let sweep = 2 * CGFloat.pi * (CGFloat(completedSteps) / CGFloat(Self.stepCount))
    let startAngle = -CGFloat.pi / 2
    let progressPath = UIBezierPath(arcCenter: CGPoint(x: ovalRect.midX, y: ovalRect.midY),
                                    radius: ovalW/2 - progressWidth/2,
                                    startAngle: startAngle,
                                    endAngle: startAngle + sweep,
                                    clockwise: true)
    progressLayer.frame = bounds
    progressLayer.path = progressPath.cgPath

    if !showDots {
      for dot in dotLayers { dot.isHidden = true }
      return
    }

    let dotRadius: CGFloat = 3.5
    let dotGap: CGFloat = 8
    let totalWidth = CGFloat(Self.stepCount) * (dotRadius * 2) + CGFloat(Self.stepCount - 1) * dotGap
    var dotX = cx - totalWidth/2 + dotRadius
    let dotY = cy + ovalH/2 + 20
    let done = progressColor.withAlphaComponent(0.5).cgColor
    let active = progressColor.cgColor
    let inactive = UIColor.white.withAlphaComponent(0.2).cgColor
    for (i, dot) in dotLayers.enumerated() {
      dot.isHidden = false
      dot.frame = CGRect(x: dotX - dotRadius, y: dotY - dotRadius, width: 7, height: 7)
      if i < activeStepIndex {
        dot.fillColor = done
      } else if i == activeStepIndex {
        dot.fillColor = active
      } else {
        dot.fillColor = inactive
      }
      dotX += dotRadius * 2 + dotGap
    }
  }
}
